clang/lib/Parse/ParseStmt.cpp

```diff
@@ -2427,3 +2447,47 @@ bool Parser::ParseOpenCLUnrollHintAttribute(ParsedAttributes &Attrs) {
   }
   return true;
 }
+
+bool Parser::ParseWhereClauseFact(WhereClause *WClause) {
+  SourceLocation Loc = Tok.getLocation();
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    IdentifierInfo *VarName = Tok.getIdentifierInfo();
+
+    // Consume the identifier.
+    ConsumeToken();
+    // Consume the ':' token.
+    ConsumeToken();
+
+    // Parse bounds expression.
+    ExprResult BoundsRes(ParseBoundsExpression());
+    return Actions.ActOnBoundsFact(WClause, VarName, BoundsRes,
+                                   getCurScope(), Loc);
+  }
+
+  ExprResult ExprRes(ParseExpression());
+  return Actions.ActOnRelopFact(WClause, ExprRes, Loc);
+}
+
+WhereClause *Parser::ParseWhereClause() {
+  SourceLocation Loc = Tok.getLocation();
+
+  // Consume the "_Where" token.
+  if (ExpectAndConsume(tok::kw__Where))
+    return nullptr;
+
+  WhereClause *WClause = Actions.ActOnWhereClause(Loc);
+
+  while (true) {
+    if (ParseWhereClauseFact(WClause))
+      return nullptr;
```

Comment:
If the current WhereClauseFact is invalid in some way, must we choose to discard all the valid facts parsed previously, or must we issue a warning that we discard the current and subsequent facts of the where clause while continuing further with the valid facts parsed until this point?

Reply:
We are currently discarding all facts in case there is an error while parsing any fact.